ENH: Implemented MultiIndex.searchsorted method ( GH14833)

GSAUC3 opened the pull request:
- [X] closes #14833
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

GSAUC3 committed:
sortedarray in side multi.py implemented, testing pending

GSAUC3 committed:
implemented the searchsorted() method, w.r.t issue #18433

GSAUC3 committed:
modified test_searchsorted, discarded the use of numpy.testing

GSAUC3 committed:
applying pre-commit fixes

datapythonista commented:
Thanks @GSAUC3 for the PR. Is there and issue, or has there been any discussion about this elsewhere?

GSAUC3 commented:
Hi @datapythonista, this was the issue https://github.com/pandas-dev/pandas/issues/14833 [https://github.com/pandas-dev/pandas/issues/14833](url) , against which i made the pull request.

GSAUC3 commented:
HI, I had ran pytest and pre-commit locally, is it possible to run all these test locally?

RB-zentronlabs commented:
@GSAUC3 Add a return statement in your function after `except` block. 
- check teh docstring and proper typing hints, `IndexOpsMixin` check this class, where searchsorted has a properly defined parameter structure.

GSAUC3 committed:
fixed the returned statement error

GSAUC3 committed:
solved the mypy type checking error, and implemented searchsorted under MultiIndex()

GSAUC3 committed:
Merge branch 'main' into searchsorted_branch

GSAUC3 commented:
hi, @datapythonista , I am having trouble running all these tests, locally, before committing my code, so far i have only ran the pytest, locally, and that worked, could you please, guide me, how to set up the testing environment locally, before each commit?

datapythonista commented:
pre-commit should run automatically if it's set up to work as intended. You have all the information on how to set up the development environment, run tests... in the development documentation: https://pandas.pydata.org/docs/development/index.html

RB-zentronlabs commented:
Hi, @datapythonista, this part of the code, fails, as well. 
```
=================================== FAILURES ===================================
__________________________ test_searchsorted[tuples] ___________________________
[gw0] darwin -- Python 3.10.17 /Users/runner/micromamba/envs/test/bin/python3.10
[XPASS(strict)] np.searchsorted doesn't work on pd.MultiIndex: GH 1[48](https://github.com/pandas-dev/pandas/actions/runs/15033468287/job/42250710830?pr=61435#step:5:52)33
___________________ test_searchsorted[mi-with-dt64tz-level] ____________________
[gw0] darwin -- Python 3.10.17 /Users/runner/micromamba/envs/test/bin/python3.10
[XPASS(strict)] np.searchsorted doesn't work on pd.MultiIndex: GH 14833
___________________________ test_searchsorted[multi] ___________________________
```

datapythonista commented:
> Hi, @datapythonista, this part of the error messages tells, us that searchsorted method should fail, but it is passing, am i correct?

Yes, that's correct. I guess we have an xfail for the test that should be removed.

GSAUC3 committed:
Closes the GH 14833 issue

GSAUC3 committed:
pre-commit run successful, closes issue 14833

GSAUC3 committed:
fixed incompatible return type issue, and closes issue 14833

GSAUC3 committed:
fixed typing checks; closes issue 14833

GSAUC3 commented:
Hi, @datapythonista, Could you please review the changes, and let me know if it meets the expectation? Thank you

GSAUC3 commented:
Hi @datapythonista . Thank you for your suggestions, I've addressed the feedback from earlier and the CI checks are now passing. This PR should be ready for review whenever you get a chance. Please let me know if any changes are required. Thanks again!

datapythonista commented on pandas/tests/base/test_misc.py:
```diff
@@ -147,14 +147,14 @@ def test_searchsorted(request, index_or_series_obj):
-            )
-        )
-    elif obj.dtype.kind == "c" and isinstance(obj, Index):
+    # if isinstance(obj, pd.MultiIndex):
+    #     # See gh-14833
+    #     request.applymarker(
+    #         pytest.mark.xfail(
+    #             reason="np.searchsorted doesn't work on pd.MultiIndex: GH 14833"
+    #         )
+    #     )
+    if obj.dtype.kind == "c" and isinstance(obj, Index):
```
Can you remove this instead of commenting please?

GSAUC3 replied:
Absolutely, I will do it right away.

datapythonista commented on .gitignore:
```diff
@@ -141,3 +141,5 @@ doc/source/savefig/
 ##############################
 /.pyodide-xbuildenv-*
+
+*.ipynb
```
Better to remove this. We do have some notebooks in this repo iirc.

GSAUC3 replied:
Yes, I will remove it.

datapythonista commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -1029,3 +1029,28 @@ def test_get_loc_namedtuple_behaves_like_tuple():
         assert idx.get_loc(("i5", "i6")) == 2
+
+
+def test_searchsorted():
```
I think it would be better to divide this test. Ideally, when a test fails, we want to know what's wrong just by checking the test name, and also we want to still test everything else. With a single test, the first thing that fails will make the whole test fail. You can use pytest fixtures and parametrize to not repeat too much code.

GSAUC3 replied:
Thank you for the suggestion, it is quite insightful for me. Would something like this be okay?

```python
def test_searchsorted_single():... # for single inputs

def test_searchsorted_lists():.. # list of single inputs

def test_searchsorted_invalid():... # for invalid inputs
```

datapythonista replied:
Yes, that may seem silly, but then when a test fails it's very easy to know what it fails. Ideally an assert per test, and if more than one input is asserted, then pytest.mark.parametrize to reuse a test with multiple inputs.

GSAUC3 replied:
No no, it is not at all silly.  I understand the importance of descriptive test functions. Thank you.

datapythonista reviewed (commented):
Looks good, added few comments.

You will have to add a not in the whatsnew for 3.0. You can check other PRs to see how we do it.

GSAUC3 committed:
updated whatsnew list, added more descriptive tests

GSAUC3 committed:
warning in rst file solved. closes issue 14833

GSAUC3 commented:
@datapythonista Hi, sorry for this silly doubt, but i haven't changed this file, 
```
Sphinx parallel build error:
RuntimeError: Unexpected exception in `/home/runner/work/pandas/pandas/doc/source/getting_started/comparison/comparison_with_sas.rst` line 135
```
and i did run local check after updating the `v3.x.x.rst` file, and it didn't throw any error either. 
and i have checked the line 135 in comparison_with_sas.rst i have checked with the current version in pandas repo, and the file is exact same, no line has been modified. could this be because of running pre-commit? 
but then i would be able to see the changes. but there were none.

GSAUC3 committed:
added searchsorted method in indexing.rst

datapythonista reviewed (approved):
Thanks @GSAUC3, this looks good to me now. I'll let someone else review it before merging.

Don't worry about the branch not being updated. This will be the case almost always, as soon as other PRs get merged, which is most days. The only cases this is an issue is when there are conflicts, but this is not the case here.

mroeschke commented on pandas/core/indexes/multi.py:
```diff
@@ -3778,6 +3788,99 @@ def _reorder_indexer(
+        if not value:
+            raise ValueError("searchsorted requires a non-empty value")
+
+        if not isinstance(value, (tuple, list)):
```
These types do not match what the type annotation suggest

GSAUC3 replied:
Thanks for the comment, you're absolutely right.
The current runtime check assumes value must be a tuple or list of tuples  but the type annotation allows for array-like or ExtensionArray inputs, which may not satisfy that condition.
I will update this.

mroeschke commented on pandas/core/indexes/multi.py:
```diff
@@ -3778,6 +3788,99 @@ def _reorder_indexer(
+                val = i if side == "left" else i + 1
+                result.append(np.intp(val))
+            else:
+                dtype = np.dtype(
+                    [
+                        (f"level_{i}", np.asarray(level).dtype)
+                        for i, level in enumerate(self.levels)
+                    ]
+                )
+
+                val_array = np.array([v], dtype=dtype)
+
+                pos = np.searchsorted(
+                    np.asarray(self.values, dtype=dtype),
+                    val_array,
+                    side=side,
+                    sorter=sorter,
+                )
+                result.append(np.intp(pos[0]))
```
This probably won't preserve pandas nullable extension types

GSAUC3 replied:
Hi @mroeschke, thanks for the comment.

Would it make sense to switch to using:
```python
dtype = np.dtype(
    [(f"level_{i}", level.dtype) for i, level in enumerate(self.levels)]
)
```
My concern here is that level.dtype might include extension dtypes (e.g., StringDtype, Int64Dtype) which are not supported by NumPy structured arrays and could raise a TypeError.

An alternative would be to avoid np.searchsorted entirely in the fallback path and implement a manual binary search over the MultiIndex.values, which would preserve pandas’ nullable types and avoid reliance on NumPy's structured dtypes.

Would you prefer we go with the binary search approach instead, or is there another path you'd recommend?

Thanks!

mroeschke reviewed (changes requested):
I think the proper solution would involve possibly allowing `algorithms.searchsorted` to support 2D inputs. The general `searchsorted` pattern is to dispatch to the array, so that `ExtensionArray` automatically gain support as well. However, `ExtensionArray` only support 1D data so I think progress on that front is needed first

GSAUC3 commented:
Hi @mroeschke, thank you for the review and helpful feedback.

I understand that ExtensionArray currently only supports 1D data, and making it work with 2D inputs would likely take some deeper changes.

If the long-term goal is to update algorithms.searchsorted to support 2D inputs and dispatch to the array — so that ExtensionArray can benefit automatically — I’d be happy to help with that.

Please let me know how you’d like to move forward. I’d be glad to contribute to any changes or help explore what’s needed.

GSAUC3 committed:
ensured typing hints consistency, closes issue 14833

GSAUC3 committed:
fixed mypy error, closes issue 14833

GSAUC3 committed:
binary search approach for searchsorted(), closes issue 14833

GSAUC3 commented:
Hi @mroeschke, I've made the required changes.
I had a question;  would it be appropriate to implement this using binary search?
I already have a working implementation ready, and I'm happy to push it if that's the recommended approach.
Let me know what you think!

GSAUC3 committed:
fixed the overload pyright error

GSAUC3 commented:
@datapythonista Hi, hope you are doing well, would you mind reviewing this pull request please?

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
